### ISSUE TITLE
Button Not being Pressed by InputEventScreenTouch

### ISSUE THREAD

**Issue by simoncrawford2018** [godotengine/godot#24589]
**Godot version:**
3.0.6

**OS/device including version:**
windows 10 pro / GT 1030 gpu

**Issue description:**
In the script extending button, when using _gui_input() and InputEventScreenTouch, it does not cause the button to be pressed down. Visually also the button does not go button_down.

**Steps to reproduce:**
Create a button, and in the code do:

func _gui_input():
   if event is InputEventScreenTouch and event.is_pressed():
       self.pressed = true

// Also if you connect the pressed() signal with this button, it will not trigger the signal when self.pressed = true. Visually the button does not go down and it does not trigger the Pressed() signal. HOWEVER, IF THE BUTTON IS SET TO TOGGLE MODE, IT WILL REGISTER THE BUTTON AS PRESSED CORRECTLY

**Minimal reproduction project:**
[Uploading InputEventScreenTouch Bug.zip…]()

**Comment by volzhs** [comment]
maybe you pressed "Submit new issue" button before completing to upload your sample project.

**Comment by simoncrawford2018** [comment]
> maybe you pressed "Submit new issue" button before completing to upload your sample project.

Sorry. I can try to upload again. However, keep in mind that this is for android export. If you try this project, you will have to make sure your the editor settings for export are correct and also the export settings.

**Comment by volzhs** [comment]
There is a option `Emulate Touchscreen` at **Project Settings > Display > Window > Handheld** to test touches on PC.
doesn't it help to test on PC?

**Comment by simoncrawford2018** [comment]
> There is a option `Emulate Touchscreen` at **Project Settings > Display > Window > Handheld** to test touches on PC.
> doesn't it help to test on PC?

Yes. You're right. However, I do not consider that a reliable way to test your input, because although it's emulating a touch, there are slight discrepancies which could affect the behavior. For example, after pressing and holding your finger down, if you move it at all, godot interprets that as a release, I think. It's behaving very strangely and I can't get it to work.

**Comment by volzhs** [comment]
yes. it does not guarantee to work exact same with touches, but helps while developing.
so, your issue occurs on the actual phone, not on the PC?

**Comment by simoncrawford2018** [comment]
> yes. it does not guarantee to work exact same with touches, but helps while developing.
> so, your issue occurs on the actual phone, not on the PC?

The issue occurs on my phone. It appears to work fine on my PC with emulate touchscreen on.

When I'm pressing and holding down on the screen with my phone, it does not constantly register "event.is_pressed()"

**Comment by volzhs** [comment]
the `InputEventScreenTouch` is occurred once only touch start or end.
when holding down, `InputEventScreenDrag` event occurs.

if you want to check it, I would suggest you to use a variable to save its state.
```gdscript
var is_touched = false
func _gui_input():
    if event is InputEventScreenTouch:
        is_touched = event.is_pressed()
```
does it help?

**Comment by simoncrawford2018** [comment]
> the `InputEventScreenTouch` is occurred once only touch start or end.
> when holding down, `InputEventScreenDrag` event occurs.
> 
> if you want to check it, I would suggest you to use a variable to save its state.
> 
> ```gdscript
> var is_touched = false
> func _gui_input():
>     if event is InputEventScreenTouch:
>         is_touched = event.is_pressed()
> ```
> 
> does it help?

I am not certain how to apply or test that. I'm sorry. I'll keep trying, though.

**Comment by lefay1982** [comment]
Yeah, when InputEventScreenDrag event occurs, all other Button can not pressed.
I find this issue in my game, when I touched and drag to fire. I can't change my weapon by click button in the same time.

**Comment by Calinou** [comment]
Can anyone still reproduce this on [Godot 3.2 beta4](https://godotengine.org/article/dev-snapshot-godot-3-2-beta-4)?

**Comment by creikey** [comment]
@Calinou I am having this issue in Godot 3.2 stable, where the Button isn't working on mobile with emulate touch with mouse, but it works on desktop.

**Comment by creikey** [comment]
It seems like the reason why this isn't working is because in `_gui_input`, there isn't a mouse dragged input event to signal for the button to be pressed, but it does receive an `InputEventScreenDrag` and `InputEventScreenTouch`

**Comment by creikey** [comment]
This issue should be redefined as an enhancement to the `BaseButton` class that lets it also detect the `InputEventScreenTouch` event in `_gui_input`

**Comment by ricochet1k** [comment]
This might be happening due to touch release events not being sent. Similar to #16761, #41585

**Comment by Bloodyaugust** [comment]
I am also observing this issue with Godot 3.2 on the Android platform. No buttons can be pressed at all.

This has been around for two years... I wonder, how is anyone building Android games in Godot if you can't press buttons? 🤔

**Comment by Calinou** [comment]
> This has been around for two years... I wonder, how is anyone building Android games in Godot if you can't press buttons? thinking

The TouchScreenButton node can (and should) be used for gameplay purposes. If you need to anchor it, add it as a child of a Control node.

**Comment by Bloodyaugust** [comment]
> > This has been around for two years... I wonder, how is anyone building Android games in Godot if you can't press buttons? thinking
> 
> The TouchScreenButton node can (and should) be used for gameplay purposes. If you need to anchor it, add it as a child of a Control node.

Right, but outside of gameplay, like in a traditional menu, `Button` and its children are what we should reach for, correct? Because they cannot be `pressed` by touch on Android.

**Comment by Calinou** [comment]
> Button and its children are what we should reach for, correct? They cannot be pressed by touch on Android.

They work here (tested on multiple projects). Can you upload a minimal reproduction project where you can't press buttons on Android?

**Comment by Bloodyaugust** [comment]
> > Button and its children are what we should reach for, correct? They cannot be pressed by touch on Android.
> 
> They work here (tested on multiple projects like [this one](https://github.com/Calinou/platshoot/)). Can you upload a minimal reproduction project where you can't press buttons on Android?

Sure, I can get that to you.

But for clarity, is that not what this issue and the linked issues are describing? That buttons cannot be pressed on Android? Perhaps my reading comprehension skills are lacking. 😅

**Comment by Calinou** [comment]
@Bloodyaugust I don't know why OP used a script for this - buttons should be pressed (and appear to be pressed) automatically when using touch input on mobile, as **Emulate Mouse From Touch** is enabled by default. **Emulate Touch From Mouse** is not enabled by default, but that's intended to be used for development on desktop platforms anyway.

**Comment by Bloodyaugust** [comment]
> ...Buttons should be pressed (and appear to be pressed) automatically when using touch input on mobile, as **Emulate Mouse From Touch** is enabled by default.
> 
> **Emulate Touch From Mouse** is not enabled by default, but that's intended to be used for development on desktop platforms anyway.

This actually may be the crux of my issue! I definitely turned _off_ `Emulate Mouse From Touch`, because I want to be able to grab multi-touch and swipe/drag inputs. I've also turned _on_ `Emulate Touch From Mouse`, and have no issue pressing the buttons while running locally with that option enabled. Did I perhaps misunderstand the intent of the former option?

**Comment by Calinou** [comment]
> I've also turned on Emulate Touch From Mouse, and have no issue pressing the buttons while running locally with that option enabled. Did I perhaps misunderstand the intent of the former option?

I believe mouse events are still sent when **Emulate Touch From Mouse** is enabled. Therefore, even nodes that only support mouse input will still work.

**Comment by Bloodyaugust** [comment]
> I believe mouse events are still sent when **Emulate Touch From Mouse** is enabled. Therefore, even nodes that only support mouse input will still work.

Do `Button` nodes and nodes that inherit from it only support mouse input?

**Comment by Calinou** [comment]
> Do `Button` nodes and nodes that inherit from it only support mouse input?

I don't remember off the top of my head. It might be worth checking the source code: https://github.com/godotengine/godot/blob/master/scene/gui/base_button.cpp

**Comment by Bloodyaugust** [comment]
Reading the source, it seems to me that indeed there is only mouse integration for button pressing. From the OP:

> Visually the button does not go down and it does not trigger the Pressed() signal.

From my own experience, when `Emulate Touch From Mouse` is off, this is true. The `pressed` signal is not fired, and no visual element of the button is interacted with. I'm not sure if this observation and tighter scoping of the issue with worth filing a new GitHub Issue, what do you think? It does seem at least loosely related, and even if the behavior is intended, some documentation on the subject would be great. It's kind of baffling to me that your only option for interacting with menus in Godot on mobile is to emulate mouse behavior.

**Comment by Calinou** [comment]
@Bloodyaugust No need to open a new issue – I think it's the same underlying issue.

**Comment by wombatwingdings** [comment]
I have same/similar issue. I switch off "emulate mouse from touch" so that I can handle selecting objects in 3D space by tapping on them (collision shape + _input_event function). That works, but I now can't use any of the standard buttons. I will have to roll my own using TouchScreenButton, which is not ideal.

**Comment by blinkakawize1** [comment]
Hi, I created this ticket like a year or two ago. Been following the updates on this issue. In all fairness, I was pretty new to godot when I first used it. I probably could find a workaround easily now with my experience. Maybe I was just doing something wrong, or maybe it actually was a problem.

**Comment by wombatwingdings** [comment]
I think UI controls not working with touch controls is a problem.

**Comment by blinkakawize1** [comment]
> I think UI controls not working with touch controls is a problem.

I agree with you. I was just worried maybe I was wrong and I looked dumb. I am pretty sure when I look back though, it really wasn't working properly as it should.

**Comment by wombatwingdings** [comment]
While this is still a problem, I have worked around it myself by leaving "Emulate Mouse From Touch" enabled and capturing mouse events instead. I can capture the swipes I'm interested in this way. Probably not good enough if you want to capture multiple finger swipes at the same time, but I'm not doing that.

**Comment by JapanYoshi** [comment]
I’m interested in letting mobile users press multiple buttons at once, which I suspect is impossible with “Emulate Mouse From Touch”.

I think buttons not working with touch controls is a problem. Godot is supposed to make building a working product easy and hassle-free; manually coding behavior that captures touch inputs on every single button is the very definition of a hassle.

**Comment by Calinou** [comment]
> I’m interested in letting mobile users press multiple buttons at once, which I suspect is impossible with “Emulate Mouse From Touch”.

Look into using TouchScreenButton nodes for this use case (instead of Button).

**Comment by JapanYoshi** [comment]
> Look into using TouchScreenButton nodes for this use case (instead of Button).
But why? Why can't UI components be used with touch controls?

**Comment by Calinou** [comment]
@JapanYoshi UI components can be used with touch controls, but you generally don't need multi-touch functionality with those as these are mainly used for menus.

For gameplay actions, TouchScreenButton handles multi-touch out of the box.

**Comment by grayhaze** [comment]
This issue explains why I couldn't get `SplitContainer`s working on Android when `Emulate mouse from touch` was disabled. Are `Control` nodes really not useable with touch input? This seems like a massive oversight given how expansive and full-featured the UI building nodes are in Godot. Surely using multi-touch in combination with UI isn't that rare a use case? `TouchScreenButton` is obviously useful for buttons, but what about all the other `Control` nodes?

**Comment by Calinou** [comment]
> Are Control nodes really not useable with touch input?

I don't think support for native touch events was implemented in most Control nodes. It's a lot of work to do properly, and mouse emulation is suitable for most developers already.

**Comment by grayhaze** [comment]
> As said above, support for _native_ touch events was never implemented in most Control nodes. It's a lot of work to do properly, and mouse emulation is suitable for most projects already. This is especially the case as most Godot developers targeting mobile platforms are creating fairly simple GUIs.

In my case, this falls apart because I'm building a multi-platform game with multiple viewports, which needs to support both touch and mouse, including pinch zoom on touch devices. I had hoped to leverage the rich UI library offered by Godot, but it seems that in order to do this I either need to take away multitouch functionality or build my own UI controls from scratch.

I imagine I'm not alone in wanting to use a rich UI in combination with multitouch gestures, but perhaps most users aren't building for multiple platforms.

**Comment by Calinou** [comment]
@grayhaze Maybe you could give [Godot Touch Input Manager](https://godotforums.org/discussion/29595/version-2-0-0-beta-of-godot-touch-input-manager-is-out) a try?

**Comment by PureAsbestos** [comment]
I have been experiencing this issue as well in the latest version of Godot (4.2.1).
I am working on a cross-platform app that will use both GUI elements and multi-touch input in different areas of the screen.

1. I disable "Emulate Mouse From Touch" as I would like to handle touch input and mouse input separately
2. I no longer receive both touch and mouse input when the user touches the screen (good)
3. I am unable to click buttons by touching them (bad)

What I expect:
Buttons to respond to touch even when mouse emulation is disabled

What I do not expect:
Multiple buttons to respond when pressed at the same time (this is what TouchScreenButton is for)

Workarounds:
1. Leave "Emulate Mouse From Touch" enabled, and explicitly ignore mouse input on mobile devices.
(What if the user connects a Bluetooth mouse to their tablet? What if their laptop has a touchscreen?)
2. Disable "Emulate Mouse From Touch", and manually extend buttons and other GUI elements to respond to touch.
(This is time consuming and repetitive, hard to get right, and liable to break.)

Neither is ideal.
As time goes on I believe the number of cross-platform Godot applications that will want access to both advanced custom multi-touch controls and the power of Godot's built-in GUI components will only increase, so this is probably worth investing in. Since the behavior of controls when "Emulate Mouse From Touch" is enabled is already pretty much what is expected, it doesn't seem to me like very much would actually need to change in terms of internal logic.

**Comment by PureAsbestos** [comment]
As an alternative to making buttons respond to touch without "Emulate Mouse From Touch" being enabled, we could add a flag to detect whether a mouse event is emulated. Then custom controls could ignore emulated mouse events. This is the easiest to implement and least invasive option I can think of that would still allow for the full functionality that I and others are looking for.

**Comment by PureAsbestos** [comment]
Perhaps a bit off-topic, but this also has me wondering why mouse events (`InputEventMouseButton` and `InputEventMouseMotion`) are grouped into `InputEventMouse` but touch events (`InputEventScreenTouch` and `InputEventScreenDrag`) are not grouped into `InputEventScreen`

**Comment by PureAsbestos** [comment]
> a flag to detect whether a mouse event is emulated

It looks like this already exists! `device` in `InputEvent` is set to `-1` (`InputEvent::DEVICE_ID_EMULATION` internally) when the input is emulated. This seems to be true for both emulated mouse events and emulated touch events, but the docs only mention that this is the case for emulated mouse events, so that should probably be updated.

This allows for a new workaround:
- Leave "Emulate Mouse From Touch" enabled, and explicitly ignore *emulated* mouse input in custom controls by checking the device ID

Which is good enough for me.

**Comment by Mohammad9760** [comment]
I'm having a similar situation as PureAsbestos described.
I'm gonna try his workaround now to see if it works.
someone should fix this issue asap

**Comment by Mohammad9760** [comment]
this issue also affects the Android Editor.
if you're developing using Godot Engine on Android, and you disable 'Emulate Mouse From Touch' then the editor just won't work at all.
it is a breaking experience, should I open a new issue?

**Comment by Zireael07** [comment]
Weird, for me the editor itself works on Android but I ran into the same issue when running the project itself.

**Comment by Mohammad9760** [comment]
> Weird, for me the editor itself works on Android but I ran into the same issue when running the project itself.

hmm.
try disabling it, close the Android Editor, Open the project again.

**Comment by PureAsbestos** [comment]
> this issue also affects the Android Editor. if you're developing using Godot Engine on Android, and you disable 'Emulate Mouse From Touch' then the editor just won't work at all. it is a breaking experience, should I open a new issue?

I would say that is a separate issue, because that option should only affect the project and not the editor.

**Comment by Calinou** [comment]
Please open a dedicated issue for that. That project setting should indeed not affect the editor itself.
